unittest/src/fake/alarm/mod.rs

```diff
@@ -0,0 +1,78 @@
+//! Fake implementation of the Alarm API.
+//!
+//! Supports frequency and set_relative.
+//! Will call the upcall immediately.
+
+use core::cell::Cell;
+use core::num::Wrapping;
+use libtock_platform::{CommandReturn, ErrorCode};
+
+use crate::upcall;
+
+pub struct Alarm {
+    frequency_hz: u32,
+    now: Cell<Wrapping<u32>>,
+}
+
+impl Alarm {
+    pub fn new(frequency_hz: u32) -> std::rc::Rc<Alarm> {
+        std::rc::Rc::new(Alarm {
+            frequency_hz,
+            now: Cell::new(Wrapping(0)),
+        })
+    }
+}
+
+impl crate::fake::SyscallDriver for Alarm {
+    fn id(&self) -> u32 {
+        DRIVER_NUM
+    }
+    fn num_upcalls(&self) -> u32 {
+        1
+    }
+
+    fn command(&self, command_number: u32, argument0: u32, _argument1: u32) -> CommandReturn {
+        match command_number {
+            command::FREQUENCY => crate::command_return::success_u32(self.frequency_hz),
+            command::SET_RELATIVE => {
+                // We're not actually sleeping, just ticking the timer.
```

Comment:
@jrvanwhy just to clarify, is there a plan to eventually use additional threads to support delayed-wake in unit tests, or is the plan always to just immediately deliver callbacks?

Reply:
I think that we can support delayed-wake in unit tests without the complexity of additional threads.

My idea is to have the fake API track a fake "current time" value that the test case increments. It should schedule the upcall when the current time passes the setpoint time. In the form of pseudocode:

```rust
let fake_alarm = fake::Alarm::new(100);
let done = Cell::new(false);
let alarm = ...;
share::scope(|...| {
    alarm.set_alarm_ms(100, &done);
    yield_no_wait();
    assert_eq!(done.get(), false);     // Upcall has not been invoked.
    fake_alarm.increment_time_ms(50);
    yield_no_wait();
    assert_eq!(done.get(), false);     // Upcall still not scheduled; only 50ms has passed.
    fake_alarm.increment_time_ms(100);
    yield_no_wait();
    assert_eq!(done.get(), true);      // Upcall invoked as 150ms has passed.
});
```

Reply:
As I mention in the next line, I don't actually know how sleeping is supposed to work, so I settled for this very naive implementation. I can change it, but first I need to know what happens in edge cases, like when setting the alarm for a moment that just passed.